### ISSUE TITLE
Indexes are not working as intended

### ISSUE THREAD

**Issue by pavlexander** [aspnet/EntityFramework6#591]
There are at least 3 issues with "Index" functionality which I would like to report. I hope you don't mind having them in 1 place.

### Steps to reproduce (issue 1)
I am trying to create 2 indexes like this:

```c#
modelBuilder.Entity<MyEntity>()
    .HasIndex(p => new { p.Column1, p.Column2, p.Column3, p.Column4 })
    .HasName("ix_index1")
    .IsUnique();

modelBuilder.Entity<MyEntity>()
    .HasIndex(p => new { p.Column1, p.Column2, p.Column3 })
    .HasName("ix_index2")
    .IsUnique();
```

Upon running command add-migration InitialCreate what I get in return, is this script:

```c#
CreateTable(
    "DEV.MyEntity",
    c => new
        {
            Column1 = c.String(nullable: false, maxLength: 10),
            Column2 = c.Decimal(nullable: false, precision: 19, scale: 0),
            Column3 = c.Decimal(nullable: false, precision: 10, scale: 0),
            Column4 = c.Decimal(nullable: false, precision: 10, scale: 0),
        })
    .PrimaryKey(t => new { t.Column1, t.Column2, t.Column3 })
    .Index(t => new { t.Column1, t.Column2, t.Column3 }, unique: true, name: "ix_index2")
    .Index(t => t.Column4, unique: true, name: "ix_index1");
```

Is there a reason why of all columns in index ix_index1 only Column4 is present? The expected result is that ix_index1 is for 4 columns.

### Steps to reproduce (issue 2)

Despite having issue 1 - I was able to create index using another approach:

```c#
modelBuilder.Entity<Invoice>()
	.Property(e => e.ExternalSystemId)
	.HasColumnAnnotation(
		IndexAnnotation.AnnotationName,
		new IndexAnnotation(new[]
		{
			new IndexAttribute("ix_invoice_invoice_id_status", 1) { IsUnique = true }
		}));

modelBuilder.Entity<Invoice>()
	.Property(e => e.ClientId)
	.HasColumnAnnotation(
		IndexAnnotation.AnnotationName,
		new IndexAnnotation(new[]
		{
			new IndexAttribute("ix_invoice_invoice_id_status", 2) { IsUnique = true }
		}));

modelBuilder.Entity<Invoice>()
	.Property(e => e.InvoiceId)
	.HasColumnAnnotation(
		IndexAnnotation.AnnotationName,
		new IndexAnnotation(new[]
		{
			new IndexAttribute("ix_invoice_invoice_id_status", 3) { IsUnique = true }
		}));

modelBuilder.Entity<Invoice>()
	.Property(e => e.Status)
	.HasColumnAnnotation(
		IndexAnnotation.AnnotationName,
		new IndexAnnotation(new[]
		{
			new IndexAttribute("ix_invoice_invoice_id_status", 4) { IsUnique = true }
		}));
```

Which creates following migration script:

```c#
CreateTable(
	"DEV4.Invoice",
	c => new
		{
			EXTERNAL_SYSTEM_ID = c.String(nullable: false, maxLength: 10),
			CLIENT_ID = c.Decimal(nullable: false, precision: 19, scale: 0),
			INVOICE_ID = c.Decimal(nullable: false, precision: 10, scale: 0),
			STATUS = c.Decimal(nullable: false, precision: 10, scale: 0),
		})
	.PrimaryKey(t => new { t.EXTERNAL_SYSTEM_ID, t.CLIENT_ID, t.INVOICE_ID })
	.ForeignKey("DEV4.Table2", t => t.EXTERNAL_SYSTEM_ID, cascadeDelete: true)
	.Index(t => new { t.EXTERNAL_SYSTEM_ID, t.CLIENT_ID, t.INVOICE_ID, t.STATUS }, unique: true, name: "ix_invoice_invoice_id_status");
```

Despite having set the name for the index "ix_invoice_invoice_id_status" - in the SQL I get following script (command: update-database -Verbose):

```c#
create table "DEV4"."Table2"
(
    "EXTERNAL_SYSTEM_ID" nvarchar2(10) not null, 
    "SYSTEM_NAME" nclob not null,
    constraint "PK_external_system__1133238746" primary key ("EXTERNAL_SYSTEM_ID")
) segment creation immediate
create table "DEV4"."Invoice"
(
    "EXTERNAL_SYSTEM_ID" nvarchar2(10) not null, 
    "CLIENT_ID" number(19, 0) not null, 
    "INVOICE_ID" number(10, 0) not null, 
    "STATUS" number(10, 0) not null,
    constraint "PK_Invoice" primary key ("EXTERNAL_SYSTEM_ID", "CLIENT_ID", "INVOICE_ID")
) segment creation immediate
begin
  execute immediate
  'create unique index "DEV4"."IX_Invoice_EXTERNAL__580661328" on "DEV4"."Invoice" ("EXTERNAL_SYSTEM_ID", "CLIENT_ID", "INVOICE_ID")';
exception
  when others then
    if sqlcode <> -1408 then
      raise;
    end if;
end;
begin
  execute immediate
  'create unique index "DEV4"."IX_Invoice_EXTERNAL__749333062" on "DEV4"."Invoice" ("EXTERNAL_SYSTEM_ID", "CLIENT_ID", "INVOICE_ID", "STATUS")';
exception
  when others then
    if sqlcode <> -1408 then
      raise;
    end if;
end;
```

So instead of having an index generated with name "ix_invoice_invoice_id_status" - I get "IX_Invoice_EXTERNAL__749333062". Why?

I also tried setting index name like this:

```C#
            CreateIndex("DEV4.Invoice",
                new[] { "EXTERNAL_SYSTEM_ID", "CLIENT_ID", "INVOICE_ID", "STATUS" },
                unique: true,
                name: "ix_invoice_invoice_id_status");
```

Same incorrect index name is in SQL.

### Steps to reproduce (issue 3)

This issue is continuation of issue 1 and 2, in some ways. So, for the next migration script I changed the data type of model's "CLIENT_ID" from LONG to STRING. Therefore, in new migration file - old index is being dropped and new index is being re-created, like so:

```C#
public override void Up()
{
	// ..
	DropIndex("DEV4.Invoice", "ix_invoice_invoice_id_status");
	// ..
	DropPrimaryKey("DEV4.Invoice");
	// ..
	AlterColumn("DEV4.Invoice", "CLIENT_ID", c => c.String(nullable: false, maxLength: 100));
	// ..
	AddPrimaryKey("DEV4.Invoice", new[] { "EXTERNAL_SYSTEM_ID", "CLIENT_ID", "INVOICE_ID" });
	// .. 
	CreateIndex("DEV4.Invoice", new[] { "EXTERNAL_SYSTEM_ID", "CLIENT_ID", "INVOICE_ID", "STATUS" }, unique: true, name: "ix_invoice_invoice_id_status");
	// ..
}
```

I get following exception:

```
begin
  execute immediate
  'create unique index "DEV4"."IX_Invoice_EXTERNAL__749333062" on "DEV4"."Invoice" ("EXTERNAL_SYSTEM_ID", "CLIENT_ID", "INVOICE_ID", "STATUS")';
exception
  when others then
    if sqlcode <> -1408 then
      raise;
    end if;
end;
Oracle.ManagedDataAccess.Client.OracleException (0x80004005): ORA-00955: name is already used by an existing object
```

As the error says - there already is an index with name "IX_Invoice_EXTERNAL__749333062". Reason why it exists - is because it was not dropped to begin with! If we scroll up the execution log up a bit, we will then see:

```
begin
  execute immediate
  'drop index "DEV4"."IX_Invoice_"';
exception
  when others then
    if sqlcode <> -1418 then
      raise;
    end if;
end;
```

EntityFramework generated a script which is dropping an index with name that never existed to begin with. Therefore, when a next index "IX_Invoice_EXTERNAL__749333062" is created we get an error that index already exists.

I consider this issue to be continuation of issue 2, because, if the names of the indexes were correct with to begin with - I wouldn't have faced issue 3, probably, though, there is still points to consider, for example, why was the drop index command truncated? Etc.

I am leaving analysis up to you. Is there a planned release date for EF 6.3.? As of this moment we can not use Indexes due to issues described above.

### Further technical details
EF version: 6.2.0
Database Provider: Oracle.ManagedDataAccess.EntityFramework
Operating system: Windows 10 x64
IDE: Visual Studio 2017 15.7.5 Pro

**Comment by phkelley** [comment]
@pavlexander - I hit one of the same problems that you did (# 1 of 3). I've raised a PR which fixes it - #677.

**Comment by ajcvickers** [comment]
Closing as duplicate.
